File: problems-originals/3301-3400/3310_remove-methods-from-project/solution.py
from typing import List


class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(n)]
        for a, b in invocations:
            graph[a].append(b)
        # Iterative DFS from k: a 10^5-long invocation chain would overflow
        # the recursion stack under the judged limits.
        suspicious = [False] * n
        suspicious[k] = True
        stack = [k]
        while stack:
            node = stack.pop()
            for nxt in graph[node]:
                if not suspicious[nxt]:
                    suspicious[nxt] = True
                    stack.append(nxt)
        # The group may only be removed when no outside method invokes
        # into it; otherwise nothing is removed at all.
        for a, b in invocations:
            if not suspicious[a] and suspicious[b]:
                return list(range(n))
        return [node for node in range(n) if not suspicious[node]]

File: problems-originals/3301-3400/3310_remove-methods-from-project/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @param {number[][]} invocations
 * @return {number[]}
 */
var remainingMethods = function (n, k, invocations) {
    const graph = Array.from({ length: n }, () => []);
    for (const [a, b] of invocations) {
        graph[a].push(b);
    }
    // Iterative DFS from k: a 10^5-long invocation chain would overflow
    // the recursion stack under the judged limits.
    const suspicious = new Array(n).fill(false);
    suspicious[k] = true;
    const stack = [k];
    while (stack.length > 0) {
        const node = stack.pop();
        for (const nxt of graph[node]) {
            if (!suspicious[nxt]) {
                suspicious[nxt] = true;
                stack.push(nxt);
            }
        }
    }
    // The group may only be removed when no outside method invokes
    // into it; otherwise nothing is removed at all.
    for (const [a, b] of invocations) {
        if (!suspicious[a] && suspicious[b]) return Array.from({ length: n }, (_, i) => i);
    }
    const remaining = [];
    for (let node = 0; node < n; node++) {
        if (!suspicious[node]) remaining.push(node);
    }
    return remaining;
};
